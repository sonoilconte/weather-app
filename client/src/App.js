import React from 'react';
import Search from './components/Search';
import WeatherDayList from './components/WeatherDayList';
import axios from 'axios';

class App extends React.Component {

	constructor() {
		super();
		this.state = {
			searchTerm: '',
			statusMessage: '',
			weatherDays: [],
		}
	}

	onSearchTermChange = (event) => {
		console.log(event.target.value);
		this.setState({ searchTerm: event.target.value });
	}

	onSearchTermSubmit = async (event) => {
		event.preventDefault();
		let statusMessage;
		let days, city, state, country;
		try {
			({
				data: {
					days,
					city,
					state,
					country,
				}
			} = await axios.get(`/api?location=${this.state.searchTerm}`));
			statusMessage = (days && days.length)
				? `Showing results for ${city}, ${state}, ${country}`
				: 'No results found';
		} catch (err) {
			console.error(err);
			statusMessage = 'Sorry, something went wrong!'
			days = [];
		}
		this.setState({
			weatherDays: days,
			statusMessage,
		});
	}

	render() {
		return (
		  <div className="App">
				<h1>Weather Forecast</h1>
				<Search
					searchTerm={this.searchTerm}
					onSearchTermChange={this.onSearchTermChange}
					onSearchTermSubmit={this.onSearchTermSubmit}
				/>
				<WeatherDayList
					statusMessage={this.state.statusMessage}
					weatherDays={this.state.weatherDays}
				/>
		  </div>
		);
	}	
}

export default App;
