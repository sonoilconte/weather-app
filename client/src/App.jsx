import React from 'react';
import Search from './components/Search';
import WeatherDayList from './components/WeatherDayList';

class App extends React.Component {
    constructor() {
        super();
        this.state = {
            searchTerm: '',
            statusMessage: '',
            weatherDays: [],
        };
    }

    onSearchTermChange = (event) => {
        this.setState({ searchTerm: event.target.value });
    };

    onSearchTermSubmit = async (event) => {
        event.preventDefault();
        let statusMessage;
        let days;
        let city;
        let state;
        let country;
        const { searchTerm } = this.state;
        try {
            const response = await fetch(`/api?location=${searchTerm}`);
            ({
                days,
                city,
                state,
                country,
            } = await response.json());
            statusMessage = (days && days.length)
                ? `Showing results for ${city}, ${state}, ${country}`
                : 'No results found';
        } catch (err) {
            console.error(err);
            statusMessage = 'Sorry, something went wrong!';
            days = [];
        }
        this.setState({
            weatherDays: days,
            statusMessage,
        });
    };

    render() {
        const {
            weatherDays,
            statusMessage,
        } = this.state;
        return (
            <div className="container">
                <div className="row justify-content-center">
                    <div className="col-md-6">
                        <h1 className="display-4">Weather Forecast</h1>
                    </div>
                </div>
                <Search
                    onSearchTermChange={this.onSearchTermChange}
                    onSearchTermSubmit={this.onSearchTermSubmit}
                />
                <WeatherDayList
                    statusMessage={statusMessage}
                    weatherDays={weatherDays}
                />
            </div>
        );
    }
}

export default App;
