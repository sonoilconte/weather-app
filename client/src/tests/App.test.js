import { render, screen, fireEvent, waitFor, getByText } from '@testing-library/react';
import renderer from 'react-test-renderer';
import App from '../App';

const mockData = {
	city: 'Denver',
	state: 'Colorado',
	country: 'United States',
	days: [
		{
			date: '2022-10-15',
			high: 62.7,
			low: 36.5
		},
		{
			date: '2022-10-16',
			high: 56.7,
			low: 42.6
		},
		{
			date: '2022-10-17',
			high: 63.9,
			low: 33.9
		},
	]
};

beforeEach(() => {
	jest.spyOn(global, 'fetch').mockResolvedValue({
		json: jest.fn().mockResolvedValue(mockData)
	});
});

afterEach(() => {
	jest.restoreAllMocks();
});

test('renders Weather text', () => {
  render(<App />);
  const text = screen.getByText(/Weather/i);
  expect(text).toBeInTheDocument();
});


it('renders App component correctly in its default state', () => {
	const tree = renderer.create(<App />)
		.toJSON();
	expect(tree).toMatchSnapshot();
});


it('has the expected data after searching for a term and clicking submit', async () => {
	render(<App />);
	const searchField = screen.getByTestId('searchField');
	// Confirm empty search field
	expect(searchField.value).toBe('');
	fireEvent.change(searchField, { target: { value: 'Denver' }});
	expect(searchField.value).toBe('Denver')

	const submitBtn = screen.getByTestId('submitBtn');
	fireEvent.click(submitBtn);
	
	// Confirm successful status message
	await screen.findByText('Showing results for Denver, Colorado, United States');
	// Confirm there are 3 div elements corresponding to the expected rendered data
	const divs = screen.queryAllByText(/2022/i);
	expect(divs).toHaveLength(3);
	// Search field should still have Denver after render
	expect(searchField.value).toBe('Denver');
});