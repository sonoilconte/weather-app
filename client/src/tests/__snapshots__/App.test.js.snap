// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders App component correctly 1`] = `
<div
  className="container"
>
  <div
    className="row justify-content-center"
  >
    <div
      className="col-md-6"
    >
      <h1
        className="display-4"
      >
        Weather Forecast
      </h1>
    </div>
  </div>
  <div
    className="row justify-content-center"
  >
    <div
      className="col-md-6"
    >
      <form
        onSubmit={[Function]}
      >
        <input
          onChange={[Function]}
          placeholder="Enter a city or zip code"
        />
        <input
          className="btn btn-primary"
          type="submit"
        />
      </form>
    </div>
  </div>
  <div
    className="row justify-content-center"
  >
    <div
      className="col-md-6"
    >
      <div
        className="status-message"
      />
    </div>
  </div>
</div>
`;

exports[`renders App component correctly in its default state 1`] = `
<div
  className="container"
>
  <div
    className="row justify-content-center"
  >
    <div
      className="col-md-6"
    >
      <h1
        className="display-4"
      >
        Weather Forecast
      </h1>
    </div>
  </div>
  <div
    className="row justify-content-center"
  >
    <div
      className="col-md-6"
    >
      <form
        onSubmit={[Function]}
      >
        <input
          data-testid="searchField"
          onChange={[Function]}
          placeholder="Enter a city or zip code"
        />
        <input
          className="btn btn-primary"
          data-testid="submitBtn"
          type="submit"
        />
      </form>
    </div>
  </div>
  <div
    className="row justify-content-center"
  >
    <div
      className="col-md-6"
    >
      <div
        className="status-message"
      />
    </div>
  </div>
</div>
`;
